src/parquet/schema.cc

```diff
@@ -620,6 +620,15 @@ const ColumnDescriptor* SchemaDescriptor::Column(int i) const {
   return &leaves_[i];
 }
 
+int SchemaDescriptor::ColumnIndex(const NodePtr& node) const {
```

Comment:
@advancedxy well, the shared_ptr itself is mutable because of the refcounting, but the `Node` object itself can be immutable just the same as a `ColumnPath` object. And imho it seems that what matters anyways in terms of unordered_map is that they are hashable and comparable which should be defined on the object and not the `shared_ptr`.
Another option is to use `std::string` for the key in conjunction with `ColumnPath::ToDotString`. But I'm guessing this will practically be slower than the O(n) array pass because of all the string manipulation.
Another possibility is to define the compare+hash over `std::vector<std::string>` and use `ColumnPath::ToDotVector`, but I don't see any major benefit over doing it with `ColumnPath`.